dev/com.ibm.ws.webcontainer.servlet.3.1/src/com/ibm/ws/webcontainer31/upgrade/H2HandlerImpl.java

```diff
@@ -51,16 +51,25 @@ public boolean isH2Request(HttpInboundConnection hic, ServletRequest request) th
         // Retrieve the needed header values for this request
         Map<String, String> h2Headers = null;
         HttpServletRequest hsrt = (HttpServletRequest) request;
-        Enumeration<String> headerNames = hsrt.getHeaderNames();
-        while (headerNames.hasMoreElements()) {
-            String headerName = headerNames.nextElement();
-            if (CONSTANT_connection.equalsIgnoreCase(headerName) || CONSTANT_upgrade.equalsIgnoreCase(headerName)) {
-                if (h2Headers == null) {
-                    h2Headers = new HashMap<String, String>();
+
+        Enumeration<String> upgradeHeaders = hsrt.getHeaders(CONSTANT_upgrade);
+        if(upgradeHeaders.hasMoreElements()) {
+            Enumeration<String> connectionHeaders = hsrt.getHeaders(CONSTANT_connection);
+            if(connectionHeaders.hasMoreElements()) {
+                h2Headers = new HashMap<String, String>();
+                while(connectionHeaders.hasMoreElements()) {
+                    h2Headers.put(CONSTANT_connection,connectionHeaders.nextElement());
+                }
+                while(upgradeHeaders.hasMoreElements()) {
+                    h2Headers.put(CONSTANT_upgrade, upgradeHeaders.nextElement());
                 }
-                h2Headers.put(headerName, hsrt.getHeader(headerName));
+            } else {
+                return false;  // connection header not found
             }
+        } else {
+            return false;  // upgrade header not found
         }
```

Comment:
```suggestion
        String connectionValue = hsrt.getHeader(CONSTANT_connection);
        String upgradeValue = hsrt.getHeader(CONSTANT_upgrade);
        if (connectionValue != null) {
            h2Headers = new HashMap<>();
            h2Headers.put(CONSTANT_connection, connectionValue);
        }
        if (upgradeValue != null) {
            if (h2Headers == null) {
                h2Headers = new HashMap<>();
            }
            h2Headers.put(CONSTANT_upgrade, upgradeValue);
        }
```

I suggest the above and then still call the isHTTP2UpgradeRequest like it was doing instead of returning false.

Reply:
I'm OK with still calling isHTTP2UpgradeRequest, was uneasy about preemptively returning false, as discussed in slack.

But that check only passes if at least both an upgrade header and connection header ... if either of those is missing, then it isn't an upgrade request.

The connection header is present with the mp-prim-ping workload that we are using for measurements in this activity, while upgrade header is not. My suspicion is that connection header may be commonly (always?) present, and upgrade header may be relatively less likely.

Based  on the above, I'd prefer to check for upgrade header first, and only check for connection header if the upgrade header is present - if not, just drop through and send the empty map to isHTTP2UpgradeRequest. This would save the pathlength to (pointlessly) retrieve the connection header, create the map, push the connection header into it, have checkIfUpgradeHeaders pull the connection header out of the map, find that the upgrade header is not present, and then return 'false'.